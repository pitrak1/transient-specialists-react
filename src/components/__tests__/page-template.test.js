import React from 'react'
import PageTemplate from '../page-template'
import { shallow } from 'enzyme'
import { Table } from '@instructure/ui-table'
import { TextInput } from '@instructure/ui-text-input'
import sinon from 'sinon'

describe('PageTemplate', () => {
  let ascending
  const columns = [
    { label: 'Serial Number', key: 'serialNumber' },
    { label: 'OEM Name', key: 'oemName' },
    { label: 'Model Name', key: 'modelName' },
    { label: 'Type Name', key: 'typeName' },
  ]
  let data
  const nameLink = 'equipment'
  const namePlural = 'Equipment'
  const nameSingular = 'Equipment'
  let onChange
  let searchValue
  let sortBy

  beforeEach(() => {
    ascending = true
    data = [
      {
        id: 1,
        serialNumber: 'Equipment a 1',
        oemName: 'OEM c 7',
        modelName: 'Model b 4',
        typeName: 'Type a 10',
      },
      {
        id: 2,
        serialNumber: 'Equipment b 2',
        oemName: 'OEM b 8',
        modelName: 'Model a 5',
        typeName: 'Type c 11',
      },
      {
        id: 3,
        serialNumber: 'Equipment c 3',
        oemName: 'OEM a 9',
        modelName: 'Model c 6',
        typeName: 'Type b 12',
      },
    ]
    onChange = sinon.stub()
    searchValue = ''
    sortBy = 'serialNumber'
  })

  const render = () =>
    shallow(
      <PageTemplate
        ascending={ascending}
        columns={columns}
        data={data}
        nameLink={nameLink}
        namePlural={namePlural}
        nameSingular={nameSingular}
        onChange={onChange}
        searchValue={searchValue}
        sortBy={sortBy}
      />,
    )

  const setSearchValue = (node, value) => {
    node
      .find(TextInput)
      .props()
      .onChange(null, value)
  }

  const setSort = (node, colNumber, colKey) =>
    node
      .find(Table.ColHeader)
      .at(colNumber - 1)
      .props()
      .onRequestSort(null, { id: colKey })

  it('shows all data if search is blank', () => {
    const node = render()
    expect(node).toMatchSnapshot()
  })

  it('hides data if no cell starts with the search value', () => {
    searchValue = 'Equipment a'
    const node = render()
    expect(node).toMatchSnapshot()
  })

  it('hides all data if no rows have cells starting with the search value', () => {
    searchValue = 'asdf'
    const node = render()
    expect(node).toMatchSnapshot()
  })

  it('searches with case insensitivity', () => {
    searchValue = 'TYPE A'
    const node = render()
    expect(node).toMatchSnapshot()
  })

  it('calls onChange with ascending true and sortBy column when clicking nonsorted column header', () => {
    const node = render()
    setSort(node, 3, 'modelName')
    expect(onChange.firstCall.args[0]).toEqual({
      ascending: true,
      sortBy: 'modelName',
    })
  })

  it('calls onChange with ascending negated when clicking sorted column header', () => {
    const node = render()
    setSort(node, 3, 'serialNumber')
    expect(onChange.firstCall.args[0]).toEqual({ ascending: false })
  })

  it('calls onChange with searchValue and value when changing the search', () => {
    const node = render()
    setSearchValue(node, 'Equipment 1')
    expect(onChange.firstCall.args[0]).toEqual({ searchValue: 'Equipment 1' })
  })
})
