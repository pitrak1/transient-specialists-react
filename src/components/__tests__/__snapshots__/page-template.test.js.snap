// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageTemplate hides all data if no rows have cells starting with the search value 1`] = `
<div>
  <Flex
    as="span"
    direction="row"
    elementRef={[Function]}
    inline={false}
    justifyItems="start"
    visualDebug={false}
    wrapItems={false}
  >
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={true}
      shrink={true}
    >
      <Heading
        border="none"
        color="inherit"
        ellipsis={false}
        level="h1"
        margin="medium"
      >
        Equipment
      </Heading>
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      margin="small"
      shrink={false}
    >
      <TextInput
        display="block"
        inputContainerRef={[Function]}
        inputRef={[Function]}
        interaction="enabled"
        isRequired={false}
        messages={Array []}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        renderLabel="Search"
        size="medium"
        textAlign="start"
        type="text"
        value="asdf"
      />
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      shrink={false}
    >
      <Link
        to="/equipment/create"
      >
        Add 
        Equipment
      </Link>
    </FlexItem>
  </Flex>
  <Table
    caption="Equipment"
    hover={true}
    layout="auto"
  >
    <Head>
      <Row>
        <ColHeader
          id="serialNumber"
          key="serialNumber"
          onRequestSort={[Function]}
          sortDirection="ascending"
          textAlign="start"
        >
          Serial Number
        </ColHeader>
        <ColHeader
          id="oemName"
          key="oemName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          OEM Name
        </ColHeader>
        <ColHeader
          id="modelName"
          key="modelName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Model Name
        </ColHeader>
        <ColHeader
          id="typeName"
          key="typeName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Type Name
        </ColHeader>
        <ColHeader
          id="Details"
          sortDirection="none"
          textAlign="start"
        />
      </Row>
    </Head>
    <Body />
  </Table>
</div>
`;

exports[`PageTemplate hides data if no cell starts with the search value 1`] = `
<div>
  <Flex
    as="span"
    direction="row"
    elementRef={[Function]}
    inline={false}
    justifyItems="start"
    visualDebug={false}
    wrapItems={false}
  >
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={true}
      shrink={true}
    >
      <Heading
        border="none"
        color="inherit"
        ellipsis={false}
        level="h1"
        margin="medium"
      >
        Equipment
      </Heading>
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      margin="small"
      shrink={false}
    >
      <TextInput
        display="block"
        inputContainerRef={[Function]}
        inputRef={[Function]}
        interaction="enabled"
        isRequired={false}
        messages={Array []}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        renderLabel="Search"
        size="medium"
        textAlign="start"
        type="text"
        value="Equipment a"
      />
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      shrink={false}
    >
      <Link
        to="/equipment/create"
      >
        Add 
        Equipment
      </Link>
    </FlexItem>
  </Flex>
  <Table
    caption="Equipment"
    hover={true}
    layout="auto"
  >
    <Head>
      <Row>
        <ColHeader
          id="serialNumber"
          key="serialNumber"
          onRequestSort={[Function]}
          sortDirection="ascending"
          textAlign="start"
        >
          Serial Number
        </ColHeader>
        <ColHeader
          id="oemName"
          key="oemName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          OEM Name
        </ColHeader>
        <ColHeader
          id="modelName"
          key="modelName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Model Name
        </ColHeader>
        <ColHeader
          id="typeName"
          key="typeName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Type Name
        </ColHeader>
        <ColHeader
          id="Details"
          sortDirection="none"
          textAlign="start"
        />
      </Row>
    </Head>
    <Body>
      <Row
        key="1"
      >
        <Cell
          key="serialNumber"
          textAlign="start"
        >
          Equipment a 1
        </Cell>
        <Cell
          key="oemName"
          textAlign="start"
        >
          OEM c 7
        </Cell>
        <Cell
          key="modelName"
          textAlign="start"
        >
          Model b 4
        </Cell>
        <Cell
          key="typeName"
          textAlign="start"
        >
          Type a 10
        </Cell>
        <Cell
          textAlign="start"
        >
          <Link
            to="/equipment/1"
          >
            Details
          </Link>
        </Cell>
      </Row>
    </Body>
  </Table>
</div>
`;

exports[`PageTemplate searches with case insensitivity 1`] = `
<div>
  <Flex
    as="span"
    direction="row"
    elementRef={[Function]}
    inline={false}
    justifyItems="start"
    visualDebug={false}
    wrapItems={false}
  >
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={true}
      shrink={true}
    >
      <Heading
        border="none"
        color="inherit"
        ellipsis={false}
        level="h1"
        margin="medium"
      >
        Equipment
      </Heading>
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      margin="small"
      shrink={false}
    >
      <TextInput
        display="block"
        inputContainerRef={[Function]}
        inputRef={[Function]}
        interaction="enabled"
        isRequired={false}
        messages={Array []}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        renderLabel="Search"
        size="medium"
        textAlign="start"
        type="text"
        value="TYPE A"
      />
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      shrink={false}
    >
      <Link
        to="/equipment/create"
      >
        Add 
        Equipment
      </Link>
    </FlexItem>
  </Flex>
  <Table
    caption="Equipment"
    hover={true}
    layout="auto"
  >
    <Head>
      <Row>
        <ColHeader
          id="serialNumber"
          key="serialNumber"
          onRequestSort={[Function]}
          sortDirection="ascending"
          textAlign="start"
        >
          Serial Number
        </ColHeader>
        <ColHeader
          id="oemName"
          key="oemName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          OEM Name
        </ColHeader>
        <ColHeader
          id="modelName"
          key="modelName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Model Name
        </ColHeader>
        <ColHeader
          id="typeName"
          key="typeName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Type Name
        </ColHeader>
        <ColHeader
          id="Details"
          sortDirection="none"
          textAlign="start"
        />
      </Row>
    </Head>
    <Body>
      <Row
        key="1"
      >
        <Cell
          key="serialNumber"
          textAlign="start"
        >
          Equipment a 1
        </Cell>
        <Cell
          key="oemName"
          textAlign="start"
        >
          OEM c 7
        </Cell>
        <Cell
          key="modelName"
          textAlign="start"
        >
          Model b 4
        </Cell>
        <Cell
          key="typeName"
          textAlign="start"
        >
          Type a 10
        </Cell>
        <Cell
          textAlign="start"
        >
          <Link
            to="/equipment/1"
          >
            Details
          </Link>
        </Cell>
      </Row>
    </Body>
  </Table>
</div>
`;

exports[`PageTemplate shows all data if search is blank 1`] = `
<div>
  <Flex
    as="span"
    direction="row"
    elementRef={[Function]}
    inline={false}
    justifyItems="start"
    visualDebug={false}
    wrapItems={false}
  >
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={true}
      shrink={true}
    >
      <Heading
        border="none"
        color="inherit"
        ellipsis={false}
        level="h1"
        margin="medium"
      >
        Equipment
      </Heading>
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      margin="small"
      shrink={false}
    >
      <TextInput
        display="block"
        inputContainerRef={[Function]}
        inputRef={[Function]}
        interaction="enabled"
        isRequired={false}
        messages={Array []}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        renderLabel="Search"
        size="medium"
        textAlign="start"
        type="text"
        value=""
      />
    </FlexItem>
    <FlexItem
      as="span"
      elementRef={[Function]}
      grow={false}
      shrink={false}
    >
      <Link
        to="/equipment/create"
      >
        Add 
        Equipment
      </Link>
    </FlexItem>
  </Flex>
  <Table
    caption="Equipment"
    hover={true}
    layout="auto"
  >
    <Head>
      <Row>
        <ColHeader
          id="serialNumber"
          key="serialNumber"
          onRequestSort={[Function]}
          sortDirection="ascending"
          textAlign="start"
        >
          Serial Number
        </ColHeader>
        <ColHeader
          id="oemName"
          key="oemName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          OEM Name
        </ColHeader>
        <ColHeader
          id="modelName"
          key="modelName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Model Name
        </ColHeader>
        <ColHeader
          id="typeName"
          key="typeName"
          onRequestSort={[Function]}
          sortDirection="none"
          textAlign="start"
        >
          Type Name
        </ColHeader>
        <ColHeader
          id="Details"
          sortDirection="none"
          textAlign="start"
        />
      </Row>
    </Head>
    <Body>
      <Row
        key="1"
      >
        <Cell
          key="serialNumber"
          textAlign="start"
        >
          Equipment a 1
        </Cell>
        <Cell
          key="oemName"
          textAlign="start"
        >
          OEM c 7
        </Cell>
        <Cell
          key="modelName"
          textAlign="start"
        >
          Model b 4
        </Cell>
        <Cell
          key="typeName"
          textAlign="start"
        >
          Type a 10
        </Cell>
        <Cell
          textAlign="start"
        >
          <Link
            to="/equipment/1"
          >
            Details
          </Link>
        </Cell>
      </Row>
      <Row
        key="2"
      >
        <Cell
          key="serialNumber"
          textAlign="start"
        >
          Equipment b 2
        </Cell>
        <Cell
          key="oemName"
          textAlign="start"
        >
          OEM b 8
        </Cell>
        <Cell
          key="modelName"
          textAlign="start"
        >
          Model a 5
        </Cell>
        <Cell
          key="typeName"
          textAlign="start"
        >
          Type c 11
        </Cell>
        <Cell
          textAlign="start"
        >
          <Link
            to="/equipment/2"
          >
            Details
          </Link>
        </Cell>
      </Row>
      <Row
        key="3"
      >
        <Cell
          key="serialNumber"
          textAlign="start"
        >
          Equipment c 3
        </Cell>
        <Cell
          key="oemName"
          textAlign="start"
        >
          OEM a 9
        </Cell>
        <Cell
          key="modelName"
          textAlign="start"
        >
          Model c 6
        </Cell>
        <Cell
          key="typeName"
          textAlign="start"
        >
          Type b 12
        </Cell>
        <Cell
          textAlign="start"
        >
          <Link
            to="/equipment/3"
          >
            Details
          </Link>
        </Cell>
      </Row>
    </Body>
  </Table>
</div>
`;
